### PR TITLE
Add battery detection

### PR DESCRIPTION
**Pull request by CameronRP**
Adds the ability to read the battery sense pin and check if powered by a battery through the dbus service

### CODE REVIEW

File: attiny.go [hunk 1]
```diff
@@ -106,6 +109,31 @@ func (a *attiny) PingWatchdog() error {
 	return a.write(watchdogReg, nil)
 }
 
+func (a *attiny) checkIsOnBattery() error {
```
Comment (mjs) [inline attiny.go]: any reason not to have this also return the "is on battery" result?

File: attiny.go [hunk 2]
```diff
@@ -106,6 +109,31 @@ func (a *attiny) PingWatchdog() error {
...
+func (a *attiny) readBatteryValue() (uint16, error) {
+	if !a.voltages.Enable {
+		return 0, nil
+	}
+	l := make([]byte, 1)
+	h := make([]byte, 1)
+	if err := a.tx(l, []byte{0x20}); err != nil {
+		return 0, err
+	}
+	if err := a.tx(h, []byte{0x21}); err != nil {
```
Comment (mjs) [inline attiny.go]: better make consts for 0x20 and 0x21

File: service.go [hunk 3]
```diff
@@ -68,17 +69,48 @@ func genIntrospectable(v interface{}) introspect.Introspectable {
...
+
+// OnBattery will return true when the input voltage is higher than 5.5V
+func (s service) OnBattery() (bool, *dbus.Error) {
+	if err := s.attinyNillCheck(); err != nil {
+		return false, makeDbusError(".OnBattery", err)
+	}
+	return s.attiny.onBattery, nil
+}
+
+func (s *service) attinyNillCheck() error {
```
Comment (mjs) [inline service.go]: It's "Nil" not "Nill"

Comment (mjs) [reply service.go]: `ensureATtinyPresent` might be a better name anyway

File: _release/attiny.yaml [hunk 4]
```diff
@@ -1,2 +1,8 @@
 pi-wake-time: 20:00
 pi-sleep-time: 20:00
+
+voltages:
+  enable: false
+  no-battery: 10
+  low-battery: 369
+  full-battery: 455
```
Comment (mjs) [inline _release/attiny.yaml:8]: what are the units for these?

**Review by mjs — COMMENTED** [review]
Looks close. Just a few things.

File: _release/attiny.yaml (hunk 4 above)
Comment (CameronRP) [reply _release/attiny.yaml:8]: These don't really have nice units. They are the analog voltage reading from the battery sense pin on the ATtiny. The analog read on the ATtiny is a 10 bit value from GND to +Ve.

Comment (mjs) [reply _release/attiny.yaml:8]: Could the config file have voltages which attiny-controller or the attiny firmware itself converts to the 10 bit values?

Comment (CameronRP) [reply _release/attiny.yaml:8]: For now I don't think this is will work as we don't have good data on how to correlate the battery sense value to a voltage. When we do we can have a voltage in the config file and attiny-controller can calculate what raw value that will be.

Comment (mjs) [reply _release/attiny.yaml:8]: Ok - makes sense.

File: attiny.go [hunk 5]
```diff
@@ -109,13 +110,12 @@ func (a *attiny) PingWatchdog() error {
...
-func (a *attiny) checkIsOnBattery() error {
+func (a *attiny) checkIsOnBattery() (bool, error) {
 	batVal, err := a.readBatteryValue()
 	if err != nil {
-		return err
+		return false, err
 	}
-	a.onBattery = batVal > a.voltages.NoBattery
-	return nil
+	return batVal > a.voltages.NoBattery, nil
```
Comment (mjs) [inline attiny.go]: Is it worth caching this result? The "am I on battery status" isn't going to change once it's powered on.

Keep returning the value as it is now, but use a cached value if it's been called before.

**Review by mjs — COMMENTED** [review]
Looks great - just one more thing

**Review by mjs — APPROVED** [review]
Nice

**Pull request merged**